### PR TITLE
feature/fu-393: FileDragAndDrop directive and component

### PR DESCRIPTION
**Pull request by AndyKIron**

### CODE REVIEW

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        fromEvent(this._element.nativeElement, 'dragleave').pipe(takeUntil(this.onDestroy$)).subscribe(this.onDragLeave.bind(this));
+
+        fromEvent(this._element.nativeElement, 'dragover').pipe(takeUntil(this.onDestroy$)).subscribe(this.onDragOver.bind(this));
+
+        fromEvent(this._element.nativeElement, 'drop').pipe(takeUntil(this.onDestroy$)).subscribe(this.onDrop.bind(this));
+    }
+
+    ngAfterViewInit() {
+        const parent = this._renderer.parentNode(this._element.nativeElement);
+        this.inputElement = this._renderer.createElement('input');
```
Comment (danielbrazg) [inline projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:67]: move logic to function `createInputFileElement`

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        this._renderer.setAttribute(this.inputElement, 'name', `input_file_${this.elementIdForClick}`);
+        this._renderer.setAttribute(this.inputElement, 'id', `input_file_${this.elementIdForClick}`);
+        if (this.multipleFiles) {
+            this._renderer.setAttribute(this.inputElement, 'multiple', '');
+        }
+        if (this.acceptFiles) {
+            this._renderer.setAttribute(this.inputElement, 'accept', this.acceptFiles);
+        }
+        this._renderer.setStyle(this.inputElement, 'display', 'none');
+        this._renderer.listen(this.inputElement, 'change', event => {
```
Comment (danielbrazg) [inline projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:78]: unlisten to all events

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.component.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * for disabling file selection
+     */
+    @Input() set fileDragAndDropDisabled(value: boolean) {
+        this.isDisabled = value;
+    }
+    /**
+     * CSS class name will set to the host element on drag over state
+     */
+    @Input() set dragOverCSSClass(value: string) {
```
Comment (danielbrazg) [inline projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.component.ts:24]: remove

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        this.dragOverCSS = value;
+    }
+    /**
+     * input.file accept attribute (file select dialog only, not grad&&drop)
+     */
+    @Input() acceptFiles: string;
+    /**
+     * input.file multiple attribute (file select dialog only, not grad&&drop)
+     */
+    @Input() multipleFiles = false;
```
Comment (danielbrazg) [inline projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:39]: change the name to `multiple`

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    /**
+     * CSS class name will set to the host element on drag over state
+     */
+    @Input() set dragOverCSSClass(value: string) {
+        this.dragOverCSS = value;
+    }
+    /**
+     * input.file accept attribute (file select dialog only, not grad&&drop)
+     */
+    @Input() acceptFiles: string;
```
Comment (danielbrazg) [inline projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:35]: change the name to `accept`

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    /**
+     * element ID for initial file selection dialog by click.
+     * if not provided used click on host
+     */
+    @Input() elementIdForClick: string;
+    //
+    /**
+     * for disabling file selection
+     */
+    @Input() set fileDragAndDropDisabled(value: boolean) {
```
Comment (danielbrazg) [inline projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:23]: change the name to disabled and add a class when in disabled mode `fusion-file-drag-drop-disabled`

**Review by danielbrazg — CHANGES_REQUESTED** [review]

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    @Input() multipleFiles = false;
+    /**
+     * output event emitter (files: FileList)
+     */
+    @Output() onHandleFileDragAndDrop = new EventEmitter();
+
+    onDestroy$ = new Subject<void>();
+    inputElement: any;
+
+    private isDisabled = false;
```
[the remark this answers is not in the document]
Comment (AndyKIron) [reply projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:48]: Ok.

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    private dragOverCSS = 'fu-drag-over';
```
[the remark this answers is not in the document]
Comment (AndyKIron) [reply projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:49]: Ok.

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+ */
+@Directive({
+    selector: '[fusionFileDragAndDrop]'
+})
+export class FileDragAndDropDirective implements OnInit, AfterViewInit, OnDestroy {
+    /**
+     * element ID for initial file selection dialog by click.
+     * if not provided used click on host
+     */
+    @Input() elementIdForClick: string;
```
[the remark this answers is not in the document]
Comment (AndyKIron) [reply projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:18]: Ok.

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    /**
+     * for disabling file selection
+     */
+    @Input() set fileDragAndDropDisabled(value: boolean) {
+        this.isDisabled = value;
+    }
+    /**
+     * CSS class name will set to the host element on drag over state
+     */
+    @Input() set dragOverCSSClass(value: string) {
```
[the remark this answers is not in the document]
Comment (AndyKIron) [reply projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:29]: OK.

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+     */
+    @Input() acceptFiles: string;
+    /**
+     * input.file multiple attribute (file select dialog only, not grad&&drop)
+     */
+    @Input() multipleFiles = false;
+    /**
+     * output event emitter (files: FileList)
+     */
+    @Output() onHandleFileDragAndDrop = new EventEmitter();
```
[the remark this answers is not in the document]
Comment (AndyKIron) [reply projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:43]: Ok.

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.component.ts (hunk 3 above)
Comment (AndyKIron) [reply projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.component.ts:24]: Ok.

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts (hunk 2 above)
Comment (AndyKIron) [reply projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:78]: changed it from  this._renderer.listen to RxJs fromEvent with .pipe(takeUntil(this.onDestroy$))

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts (hunk 4 above)
Comment (AndyKIron) [reply projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:39]: OK.

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts (hunk 5 above)
Comment (AndyKIron) [reply projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:35]: OK.

File: projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts (hunk 6 above)
Comment (AndyKIron) [reply projects/fusion-ui/components/file-drag-and-drop/file-drag-and-drop.directive.ts:23]: Ok. just with `fu-file-drag-drop-disabled`

**Review by danielbrazg — APPROVED** [review]

**Pull request merged**